File: src/main/java/xyz/phanta/cbtweaker/multiblock/MultiBlockControllerBlock.java
package xyz.phanta.cbtweaker.multiblock;

import io.github.phantamanta44.libnine.util.world.WorldBlockPos;
import net.minecraft.block.Block;
import net.minecraft.block.ITileEntityProvider;
import net.minecraft.block.state.IBlockState;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.tileentity.TileEntity;
import net.minecraft.util.EnumFacing;
import net.minecraft.util.EnumHand;
import net.minecraft.util.math.BlockPos;
import net.minecraft.world.World;
import xyz.phanta.cbtweaker.CbtMod;
import xyz.phanta.cbtweaker.common.OrientableMachineBlock;
import xyz.phanta.cbtweaker.gui.CbtGuiIds;

import javax.annotation.Nullable;

public class MultiBlockControllerBlock extends OrientableMachineBlock implements ITileEntityProvider {

    private final MultiBlockType<?, ?, ?> mbType;

    public MultiBlockControllerBlock(MultiBlockType<?, ?, ?> mbType) {
        super(mbType.getBlockMaterial());
        this.mbType = mbType;
    }

    public MultiBlockType<?, ?, ?> getMultiBlockType() {
        return mbType;
    }

    @Nullable
    @Override
    public TileEntity createNewTileEntity(World world, int meta) {
        return new MultiBlockControllerTileEntity();
    }

    @Override
    public String getTranslationKey() {
        return CbtMod.MOD_ID + ".multiblock." + mbType.getId() + ".controller";
    }

    @SuppressWarnings("deprecation")
    @Override
    public void neighborChanged(IBlockState state, World world, BlockPos pos, Block block, BlockPos fromPos) {
        TileEntity tile = world.getTileEntity(pos);
        if (tile instanceof MultiBlockControllerTileEntity) {
            ((MultiBlockControllerTileEntity)tile).getRedstoneHandler().setRedstoneState(world.isBlockPowered(pos));
        }
    }

    @Override
    public boolean onBlockActivated(World world, BlockPos pos, IBlockState state, EntityPlayer player, EnumHand hand,
                                    EnumFacing facing, float hitX, float hitY, float hitZ) {
        CbtMod.INSTANCE.getGuiHandler().openGui(player, CbtGuiIds.MB_CONTROLLER, new WorldBlockPos(world, pos));
        return true;
    }

}

File: src/main/kotlin/st/evening/mc/cbtweaker/multiblock/MultiBlockControllerBlock.kt
package st.evening.mc.cbtweaker.multiblock

import net.minecraft.block.Block
import net.minecraft.block.state.IBlockState
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.util.EnumFacing
import net.minecraft.util.EnumHand
import net.minecraft.util.math.BlockPos
import net.minecraft.world.World
import st.evening.mc.cbtweaker.CbTweaker
import st.evening.mc.cbtweaker.CbtConsts
import st.evening.mc.cbtweaker.common.CustomBlockType
import st.evening.mc.cbtweaker.common.OrientableMachineBlock
import st.evening.mc.prelude.api.block.TileEntityBlock
import st.evening.mc.prelude.api.registration.TileEntityType
import st.evening.mc.prelude.api.registration.openContainer
import st.evening.mc.prelude.api.util.game.SidednessAssertion
import st.evening.mc.prelude.api.util.game.assertLogicalServer
import st.evening.mc.prelude.api.util.world.onServer
import st.evening.mc.prelude.api.util.world.useTileEntity

class MultiBlockControllerBlock private constructor(val mbType: MultiBlockType<*>) :
    OrientableMachineBlock(mbType.blockConfig.material), TileEntityBlock {
    companion object {
        fun construct(mbType: MultiBlockType<*>): MultiBlockControllerBlock =
            MultiBlockControllerBlock(mbType).also { it.init() }
    }

    override val blockType: CustomBlockType
        get() = mbType

    override fun getTileEntityType(world: World, meta: Int): TileEntityType<*> =
        CbTweaker.defns.tileEntityMultiBlockController

    override fun onBlockActivated(
        world: World, pos: BlockPos, state: IBlockState, player: EntityPlayer, hand: EnumHand,
        facing: EnumFacing, hitX: Float, hitY: Float, hitZ: Float
    ): Boolean {
        world.useTileEntity<MultiBlockControllerTileEntity>(pos) {
            if (it.handleInteraction(state, player, hand, facing, hitX, hitY, hitZ)) return true
            world.onServer {
                player.openContainer(CbTweaker.defns.containerMultiBlockController, world, pos)
            }
            return true
        }
        return false
    }

    @OptIn(SidednessAssertion::class)
    @Suppress("OVERRIDE_DEPRECATION")
    override fun neighborChanged(state: IBlockState, world: World, pos: BlockPos, block: Block, fromPos: BlockPos) {
        assertLogicalServer {
            world.useTileEntity<MultiBlockControllerTileEntity>(pos) {
                it.handleBlockUpdate(state, block, fromPos)
            }
        }
    }

    @OptIn(SidednessAssertion::class)
    override fun breakBlock(world: World, pos: BlockPos, state: IBlockState) {
        assertLogicalServer {
            world.useTileEntity<MultiBlockControllerTileEntity>(pos) {
                it.handleDestruction(state)
            }
        }
        super.breakBlock(world, pos, state)
    }

    override fun getTranslationKey(): String = "${CbtConsts.MOD_ID}.multiblock.${mbType.id}.controller"
}
